File: cpp/testsuite/ApplicationCases/SceneEditorApplication/src/model/history/SceneElementKind.cpp
#include "java/util/ArrayList.txx"
#include "vsdk/toolkit/environment/camera/Camera.h"
#include "vsdk/toolkit/environment/light/Light.h"
#include "vsdk/toolkit/environment/scene/SimpleBody.h"
#include "vsdk/toolkit/environment/scene/SimpleBodyGroup.h"
#include "vsdk/toolkit/environment/scene/SimpleScene.h"
#include "model/Scene.h"
#include "model/history/SceneElementKind.h"
#include "model/selection/SelectionSet.h"

EntityListView SceneElementKind::getList(Value kind, Scene* scene)
{
    switch ( kind ) {
      case BODY:
        return EntityListView(&scene->scene->getSimpleBodies());
      case LIGHT:
        return EntityListView(&scene->scene->getLights());
      case CAMERA:
        return EntityListView(&scene->scene->getCameras());
      default:
        return EntityListView(&scene->debugThingGroups);
    }
}

SelectionSet* SceneElementKind::getSelection(Value kind, Scene* scene)
{
    switch ( kind ) {
      case BODY:
        return scene->selectedThings;
      case LIGHT:
        return scene->selectedLights;
      case DEBUG_GROUP:
        return scene->selectedDebugThingGroups;
      default:
        return nullptr;
    }
}

void SceneElementKind::insert(Value kind, Scene* scene, int index,
                              Entity* element, bool selected)
{
    SelectionSet* selection = getSelection(kind, scene);

    if ( selection != nullptr ) {
        selection->insertElement(index, element, selected);
    }
    else {
        getList(kind, scene).insert(index, element);
    }
}

void SceneElementKind::remove(Value kind, Scene* scene, int index)
{
    SelectionSet* selection = getSelection(kind, scene);

    if ( selection != nullptr ) {
        selection->removeElement(index);
    }
    else {
        getList(kind, scene).remove(index);
    }
}

File: java/testsuite/ApplicationCases/SceneEditorApplication/src/model/history/SceneElementKind.java
package model.history;

import java.util.ArrayList;

import vsdk.toolkit.common.Entity;

import model.Scene;
import model.selection.SelectionSet;

/**
Kinds of elements of a scene whose creation and deletion are undoable: each
kind is kept in its own list of the scene, most of them with a selection set
over it.
*/
public enum SceneElementKind
{
    BODY,
    LIGHT,
    CAMERA,
    DEBUG_GROUP;

    /**
    @param scene scene holding the elements
    @return the list of the scene holding the elements of this kind
    */
    public ArrayList<? extends Entity> getList(Scene scene)
    {
        switch ( this ) {
          case BODY:
            return scene.scene.getSimpleBodies();
          case LIGHT:
            return scene.scene.getLights();
          case CAMERA:
            return scene.scene.getCameras();
          default:
            return scene.debugThingGroups;
        }
    }

    /**
    @param scene scene holding the elements
    @return the selection set over the elements of this kind, or null if this
    kind can not be selected (scene cameras)
    */
    public SelectionSet getSelection(Scene scene)
    {
        switch ( this ) {
          case BODY:
            return scene.selectedThings;
          case LIGHT:
            return scene.selectedLights;
          case DEBUG_GROUP:
            return scene.selectedDebugThingGroups;
          default:
            return null;
        }
    }

    /**
    Inserts an element in its list, keeping the selection aligned.
    @param scene scene holding the elements
    @param index position of the element in the list
    @param element element to insert
    @param selected true if the element must be selected
    */
    @SuppressWarnings("unchecked")
    public void insert(Scene scene, int index, Entity element, boolean selected)
    {
        SelectionSet selection = getSelection(scene);

        if ( selection != null ) {
            selection.insertElement(index, element, selected);
        }
        else {
            ((ArrayList<Entity>)getList(scene)).add(index, element);
        }
    }

    /**
    Removes an element from its list, keeping the selection aligned.
    @param scene scene holding the elements
    @param index position of the element in the list
    */
    public void remove(Scene scene, int index)
    {
        SelectionSet selection = getSelection(scene);

        if ( selection != null ) {
            selection.removeElement(index);
        }
        else {
            getList(scene).remove(index);
        }
    }
}
